### DISCUSSION TITLE
How the API should look like

### DISCUSSION THREAD

**Discussion by pierreaubert** [pierreaubert/spinorama#237] [Ideas]
### Basic

- get(f"/{VERSION}/speaker")
  - return list of speakers
- app.get(f"/{VERSION}/speaker/{speaker_name}"/metadata)
  - return metadata for speaker
-app.get(f"/{VERSION}/speaker/{speaker_name}"/version/{version}/measurement/{measurement_name}/format/{measurement_format})
  - return measurement for speaker
- app.get(f"/{VERSION}/speaker/{speaker_name}"/eq/{kind}/format/{eq_format})
  - return EQ for speaker
- app.get(f"/{VERSION}/compare")
  - return a comparison between 2 speakers for a measurement in format
- app.get(f"/{VERSION}/similar/{speaker_name}")
  - return a list of similar speakers

### Questions

- how to add filters?
- how to sort ?  
- for compare, we need a lot of parameters; is that better in the URL
  or as parameters?

**Comment by jeromeof** [comment]
I think you probably have the correct 'shape' to the API above, though  we need to get access to the 'code' notebook capability before we can know for sure if the API has the right, I still haven't gotten access. The way I believe it will probably work, is the 'code' notebook default to use something like pandas for storing and maculating datasets and matlibplot python libraries for rendering these datasets. The Plugin API therefore is about loading the specific dataset into pandas. 

So the user would then do filtering, sorting and possible comparisons by using English commands on the datasets loaded. I have tried a few samples already in chatGPT and it can nicely generate working notebook 'steps' from well described data and manipulate that data. 

Is the API available yet? I can possible test it for you and look at what type of chatGPT 'prompts' can be engineered for this type of data. e.g. with the chatGPT API itself you can give it instructions for how it should work in a particular session. E.g. something like "You are a Spinorama assistant and should render FR using matlibplot using 'Frequency' on the horizontal access and SPL on the vertical access' 

For comparisons it might be able to load multiple datasets and then using pandas for the comparison, so you might not need to provide the comparison data itself. e.g. https://datatofish.com/compare-values-dataframes/ 

Try something like this in chatGPT: 

"Show me some notebook code to compare 2 speaker datasets by frequency and SPL in pandas before rendering in matlibplot" 

This response is obviously just made up, but I think with correct metadata and a good API (like you are defining) the code generated should be more accurate.

**Reply by jeromeof** [reply]
Hi, just want to apologise for not getting back to look at the plugin properly yet. It seems that the rollout by OpenAI of plugin access is taking forever, but I decided to wait until I can at least properly try it before making more comments.

Not sure if it's because I am in Europe rather than America but I still do not see the plugin's within my openai account. Anyway I will keep you informed as soon as I do get access. 

FYI: I thought this YouTube video chat with Stephen Wolfram was interesting https://www.youtube.com/watch?v=uD353DeOM-4 - not necessarily about plugins but in general about people interfacing with this technology to do 'programming' tasks without being a programmer. 

My day job got very busy making a large release for the past month, but I have scheduled time in 2-3 weeks time to actually play with some of the 'local' models (cut down LLM's like alpaca etc)  and how they can be plugged directly into retrieval of local data using LongChain. Not sure if it will be interesting to you its no "plugins" and but for some use-cases I believe it will be a better approach.

**Comment by jeromeof** [comment]
Hi,
  So finally I have the beta plugin access. Do you have a "dev" or "beta" deployment so I can try the plugin configuration? 
Basically it has a simple way of adding a url (for personal testing) but obviously this code isn't live yet on the www.spinorama.org etc as I get this following message: 

<img width="989" alt="image" src="https://github.com/pierreaubert/spinorama/assets/110913/db8c31fe-5ff5-4858-82ab-471ebf1fe1d6">

Anyway, I will ping you privately on ASR also

**Comment by pierreaubert** [comment]
can you try

https://api.spinorama.org

instead?

PIerre

On Sat, May 13, 2023 at 11:26 AM jeromeof ***@***.***> wrote:

> Hi,
> So finally I have the beta plugin access. Do you have a "dev" or "beta"
> deployment so I can try the plugin configuration?
> Basically it has a simple way of adding a url (for personal testing) but
> obviously this code isn't live yet on the www.spinorama.org etc as I get
> this following message:
> [image: image]
> <https://user-images.githubusercontent.com/110913/238126130-db8c31fe-5ff5-4858-82ab-471ebf1fe1d6.png>
>
> Anyway, I will ping you privately on ASR also
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pierreaubert/spinorama/discussions/237#discussioncomment-5891148>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANBKHZUEXESOEE5GY6VFC3XF5HTPANCNFSM6AAAAAAWTUV53M>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
--
PfA

**Reply by jeromeof** [reply]
I just did - it spins for while then says the same thing:

<img width="791" alt="image" src="https://github.com/pierreaubert/spinorama/assets/110913/7880c1ca-65e0-422d-a07a-0ae37012035c">

I can properly investigate tomorrow, they only seem to have a maybe 20/30 plugins at the moment - and I can't see the "notebook" one which was the one I thought would be great for your database. But I will spend more time tomorrow investigating. 

<img width="1295" alt="image" src="https://github.com/pierreaubert/spinorama/assets/110913/665293f6-81ed-438d-8289-5c9270abd49e">

There is a separate option to connect via http://localhost etc so I might just see can I run your branch locally and connect that way and at least see if that works.

**Comment by jeromeof** [comment]
Hi Pierre,
  So I properly checked out the code yesterday but was having trouble installing all the dependencies - I will investigate a bit further later on (I have a few node projects locally and a few python) but possible some mismatch in versioning is causing some issues. 

   But I thought I would investigate a little more what might be going wrong. Checking out their later documentation for the plugin:

https://platform.openai.com/docs/plugins/getting-started/plugin-manifest

And checking the URL's that are live on your website. Possible the following is going wrong. 
First thing they require is :

ai-plugin.json in the .well-known  folder

Checking the source I can see a folder under the api with that file etc. 

But trying the current deployment on your website I get an unknown 

https://api.spinorama.org/.well-known/ai-plugin.json

Then I thought I would double check the json contents itself (as checked into the develop branch) and I see a few things that are possible wrong e.g. 
"api_spec_url": "https://www.spinorama.org/.well-known/openapi.yaml",

is pointing at www. instead of api. its posting at the ./well-known folder instead of the root and your file is openapi.yml not openapi.yaml

Also it looks like what is currently deployed is here:

https://api.spinorama.org/openapi.json

So maybe that is an older version or your deployment scripts do a yml to json conversation? 

Looking at your yml file in your develop branch - I see a couple of things that are probably wrong: e.g. 
- url: "https://www.spinorama.org/api"

When I imagine you mean https://api.spinorama.org etc

I can prepare a fork / PR for you but maybe it will be easier when you get access to the plugins yourself. They haven't given me the 'Code interpreter' access yet which is were I believe your data will be amazing (I see you use Jupyter-lab already so you know how powerful a 'notebook' 'code' that be, etc. So I just have access to the basic third party plugins and the ability to test one with a custom URL etc

**Comment by pierreaubert** [comment]
hello,

i will have a look. I will provide both json file (in .well-known and in
the api). let's see if that helps.

PIerre


On Mon, May 15, 2023 at 12:59 PM jeromeof ***@***.***> wrote:

> Hi Pierre,
> So I properly checked out the code yesterday but was having trouble
> installing all the dependencies - I will investigate a bit further later on
> (I have a few node projects locally and a few python) but possible some
> mismatch in versioning is causing some issues.
>
> But I thought I would investigate a little more what might be going wrong.
> Checking out their later documentation for the plugin:
>
> https://platform.openai.com/docs/plugins/getting-started/plugin-manifest
>
> And checking the URL's that are live on your website. Possible the
> following is going wrong.
> First thing they require is :
>
> ai-plugin.json in the .well-known folder
>
> Checking the source I can see a folder under the api with that file etc.
>
> But trying the current deployment on your website I get an unknown
>
> https://api.spinorama.org/.well-known/ai-plugin.json
>
> Then I thought I would double check the json contents itself (as checked
> into the develop branch) and I see a few things that are possible wrong e.g.
> "api_spec_url": "https://www.spinorama.org/.well-known/openapi.yaml",
>
> is pointing at www. instead of api. its posting at the ./well-known
> folder instead of the root and your file is openapi.yml not openapi.yaml
>
> Also it looks like what is currently deployed is here:
>
> https://api.spinorama.org/openapi.json
>
> So maybe that is an older version or your deployment scripts do a yml to
> json conversation?
>
> Looking at your yml file in your develop branch - I see a couple of things
> that are probably wrong: e.g.
>
>    - url: "https://www.spinorama.org/api"
>
> When I imagine you mean https://api.spinorama.org etc
>
> I can prepare a fork / PR for you but maybe it will be easier when you get
> access to the plugins yourself. They haven't given me the 'Code
> interpreter' access yet which is were I believe your data will be amazing
> (I see you use Jupyter-lab already so you know how powerful a 'notebook'
> 'code' that be, etc. So I just have access to the basic third party plugins
> and the ability to test one with a custom URL etc
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pierreaubert/spinorama/discussions/237#discussioncomment-5903945>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANBKHYYET2DLBKAXI6S5FDXGIEBZANCNFSM6AAAAAAWTUV53M>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
--
PfA

**Comment by pierreaubert** [comment]
i need to check but it may want a yaml file and not a json one

On Mon, 15 May 2023 at 13:22 Pierre Aubert ***@***.***> wrote:

> hello,
>
> i will have a look. I will provide both json file (in .well-known and in
> the api). let's see if that helps.
>
> PIerre
>
>
> On Mon, May 15, 2023 at 12:59 PM jeromeof ***@***.***>
> wrote:
>
>> Hi Pierre,
>> So I properly checked out the code yesterday but was having trouble
>> installing all the dependencies - I will investigate a bit further later on
>> (I have a few node projects locally and a few python) but possible some
>> mismatch in versioning is causing some issues.
>>
>> But I thought I would investigate a little more what might be going
>> wrong. Checking out their later documentation for the plugin:
>>
>> https://platform.openai.com/docs/plugins/getting-started/plugin-manifest
>>
>> And checking the URL's that are live on your website. Possible the
>> following is going wrong.
>> First thing they require is :
>>
>> ai-plugin.json in the .well-known folder
>>
>> Checking the source I can see a folder under the api with that file etc.
>>
>> But trying the current deployment on your website I get an unknown
>>
>> https://api.spinorama.org/.well-known/ai-plugin.json
>>
>> Then I thought I would double check the json contents itself (as checked
>> into the develop branch) and I see a few things that are possible wrong e.g.
>> "api_spec_url": "https://www.spinorama.org/.well-known/openapi.yaml",
>>
>> is pointing at www. instead of api. its posting at the ./well-known
>> folder instead of the root and your file is openapi.yml not openapi.yaml
>>
>> Also it looks like what is currently deployed is here:
>>
>> https://api.spinorama.org/openapi.json
>>
>> So maybe that is an older version or your deployment scripts do a yml to
>> json conversation?
>>
>> Looking at your yml file in your develop branch - I see a couple of
>> things that are probably wrong: e.g.
>>
>>    - url: "https://www.spinorama.org/api"
>>
>> When I imagine you mean https://api.spinorama.org etc
>>
>> I can prepare a fork / PR for you but maybe it will be easier when you
>> get access to the plugins yourself. They haven't given me the 'Code
>> interpreter' access yet which is were I believe your data will be amazing
>> (I see you use Jupyter-lab already so you know how powerful a 'notebook'
>> 'code' that be, etc. So I just have access to the basic third party plugins
>> and the ability to test one with a custom URL etc
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/pierreaubert/spinorama/discussions/237#discussioncomment-5903945>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AANBKHYYET2DLBKAXI6S5FDXGIEBZANCNFSM6AAAAAAWTUV53M>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>
>
> --
> --
> PfA
>
-- 
--
PfA

**Comment by jeromeof** [comment]
So I got access to one of the other 'plugins' the Web Browser one - and decided to see what it can do with your data and it not bad I think:  

<img width="693" alt="image" src="https://github.com/pierreaubert/spinorama/assets/110913/08b9248e-c8b4-4ef5-bade-6bb701b14bb0">

<img width="654" alt="image" src="https://github.com/pierreaubert/spinorama/assets/110913/4b30817b-928a-41e3-bda6-4451af4f8a0c">

Obviously allowing people do their own analysis (using your plugin directly) I think will be much better but it looked like the Browser plugin was able to extract the data from your website and do some basic comparisons which is sort of interesting I think

**Comment by pierreaubert** [comment]
yesterday I adapted the API to the new format (json and yaml). Possibly you
can add it to the registry? I am very busy and dont have the time now to
look at it.

Pierre

On Wed, 17 May 2023 at 01:47 jeromeof ***@***.***> wrote:

> So I got access to one of the other 'plugins' the Web Browser one - and
> decided to see what it can do with your data and it not bad I think:
> [image: image]
> <https://user-images.githubusercontent.com/110913/238795030-08b9248e-c8b4-4ef5-bade-6bb701b14bb0.png> [image:
> image]
> <https://user-images.githubusercontent.com/110913/238795540-4b30817b-928a-41e3-bda6-4451af4f8a0c.png>
>
> Obviously allowing people do their own analysis (using your plugin
> directly) I think will be much better but it looked like the Browser plugin
> was able to extract the data from your website and do some basic
> comparisons which is sort of interesting I think
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pierreaubert/spinorama/discussions/237#discussioncomment-5921696>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANBKH5RMQTJ6HJEJI45YLTXGQGYZANCNFSM6AAAAAAWTUV53M>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
--
PfA

**Comment by jeromeof** [comment]
OK - yes I can see you have deployed the right files in what I believe are the right places - but sadly it still says Still Can't find plugin and I try and register it. 

Anyway, I will investigate separately over the next few evenings and hopefully get an answer.

I completely understand you are too busy (as am I really) and really this is just an experiment so its not important, but may eventually be something which people might find useful with your data, so I will see can I figure something out separately.

**Comment by jeromeof** [comment]
OK I figured out the problem - basically their plugin registration code doesn't work on Safari - I switched to Firefox and it has correctly identified the plugin (logo is wrong etc) 

<img width="646" alt="image" src="https://github.com/pierreaubert/spinorama/assets/110913/468ba1ff-6b82-4403-a550-5b9774b4d515">

And I can see if requesting the data from the website:
<img width="792" alt="image" src="https://github.com/pierreaubert/spinorama/assets/110913/897d2ab9-8fa4-4504-ada1-1781bbf819a3">

Trying to compare some speakers I do see an error in the plugin response: 
<img width="510" alt="image" src="https://github.com/pierreaubert/spinorama/assets/110913/6c976648-7360-4da8-9137-2111db40740e">

But well done it will be great to see can be done with this when the "Code interpreter" plugin is released and this was just chatGPT itself

**Reply by pierreaubert** [reply]
I fixed a few bugs and the API is now not crashing that often. See [patch](https://github.com/pierreaubert/spinorama/commit/836208cc964476fdaae4f1cd2e55bc88423a6f91).

**Comment by pierreaubert** [comment]
Closing since this API is obsolete
